policy/policy.go

```diff
@@ -2,14 +2,90 @@
 // SPDX-License-Identifier: Apache-2.0
 package policy
 
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+// PolicyID identifies a specific policy. This is used to retrieve the policy
+// from the store.
+type PolicyID struct {
+	// TenantId  is the ID of the tenant that owns this policy.
+	TenantId string `json:"tenant_id"`
+
+	// Scheme is the name of the scheme with which this policy is associated
+	Scheme string `json:"scheme"`
+
+	// Name is the name of this policy
+	Name string `json:"name"`
+}
+
+// PolicyIDFromKey parses the specified string containing a policy store key
+// into a PolicyID.
+func PolicyIDFromStoreKey(s string) (PolicyID, error) {
+	var pid PolicyID
+
+	parts := strings.Split(s, ":")
+	if len(parts) != 3 {
+		return pid, fmt.Errorf(
+			"bad policy store key %q: want 3 :-separated parts, found %d",
+			s, len(parts),
+		)
+	}
+
+	pid.TenantId = parts[0]
+	pid.Scheme = parts[1]
+	pid.Name = parts[2]
+
+	return pid, pid.Validate()
+}
+
+func (o PolicyID) Validate() error {
+	if url.PathEscape(o.TenantId) != o.TenantId {
+		return fmt.Errorf("bad TenantId %q: must be a valid URI path segment", o.TenantId)
+	}
+
+	if url.PathEscape(o.Scheme) != o.Scheme {
+		return fmt.Errorf("bad Scheme %q: must be a valid URI path segment", o.Scheme)
+	}
+
+	if url.PathEscape(o.Name) != o.Name {
+		return fmt.Errorf("bad Name %q: must be a valid URI path segment", o.Name)
+	}
+
+	return nil
+}
+
+// StoreKey returns the key used with the policy store.
+func (o PolicyID) StoreKey() string {
+	return fmt.Sprintf("%s:%s:%s", o.TenantId, o.Scheme, o.Name)
+}
+
 // Policy allows enforcing additional constraints on top of the regular attestation schemes.
 type Policy struct {
-	// ID is used to reference the policy in the result.
-	ID string `json:"id"`
+	// ID is the identifier of this policy, unique to the store.
+	ID PolicyID `json:"-"`
 
 	// Version gets bumped every time a new policy with existing ID is added to the store.
 	Version int32 `json:"version"`
 
 	// Rules of the policy to be interpreted and execute by the policy agent.
 	Rules string `json:"rules"`
 }
+
+// VersionedName returns a string identifier for the policy which is derived
+// from its policy ID and version (and is therefore unique across versions).
+func (o Policy) VersionedName() string {
```

Comment:
To be consistent, with the convention, why should a Policy Versioned Name not comprise of
PolicyID and Version, meaning why does the Versioned Name, not include the full PolicyID, which includes Scheme as well ?

Reply:
because `VersionedName` is used to identify the policy inside the `appraisal-policy-id`, which already contains the scheme.

Reply:
If we construct a consistent name, here which includes Tenant, scheme, name and version then we can overwrite the AppriasalPolicyID from the default scheme name at init to this name completely.

This way the VersionedName becomes consistent and does not rely on who is using this?
This is a public API afterall...

Reply:
we could, however, then:
- `appraisal-policy-id`  fomat will be incosinstent beteween cases when a policy when it was applied or it wasn't
- we increase redundancy in the `appraisal-policy-id` when we start supporting multiple policies.

Reply:
Yes, I thought a bit about it, if we state that the Versioned Policy ID (appraisal policy) is a combination of
Scheme, Tenant, name and version in that order then we can start with Scheme (default) and then replace with the actual `PolicyID` without losing the consistency ?

Reply:
True, however, this doesn't solve the second problem.

Reply:
I do not necessarily understand when you state increasing redundancy specifically. Also at the most I can understand per sub-mod specific AppraisalPolicyID which will entail  adding the subModName followed by the Scheme to provide the required uniqueness, in the PolicyID construct when setting the final `appraisal-policy-id`

Reply:
You would have duplication of scheme for each policy: `policy:SCHEME:0:base:v1/SCHEME:0:extra:v1`.

Conceptually: _either_ the policy generates the entire Appraisal ID (overriding the one set in cases where there is no policy), _or_ the policy only generates the part of the Appraisal ID  that represents itself.

If the policy generates the entire thing, then we have a problem in the mulitiple policy case, as they're unaware of each other, and the  last policy would simply override the id set py the previousl policies losing information.

If the policy only generates its part of the overall ID, then  the scheme should not be included in that part because the scheme is not part of the policy (even though a policy is tied to a  particular scheme, the scheme is _not_ tied to that policy).

Reply:
> per sub-mod specific AppraisalPolicyID

This is possible in our current pipleline -- submods are handled on per-scheme basis. The pipeline has no awareness of individual submods, and the policy agent operates on the attestation result as a whole, not individual submod appraisals.

Reply:
To my knowledge  at the moment the interface does not support the multiple policy handling (per submod) at the moment.

We have two options, we construct one unique `appraisal-policy-id` from the set of policies.
This can be effectively handled internally.

OR I assume we will have an array of  `appraisal-policy-id` strings and each UpdatePolicyID will append to that existing array. So I do not see the overwrite use case, in any way.

In either case, adding Scheme to the policy is always a wise and safe thing, even if there is a bit of duplication!

Reply:
> OR I assume we will have an array of appraisal-policy-id strings and each UpdatePolicyID will append to that existing array. So I do not see the overwrite use case, in any way.

This is effectively what is happening now (execept it's a single string that is being appended to, rather than array). However if each policy's "id" is being appended to it, tha "id" does not need to include the scheme because the scheme is already present (and is the same for all policies).

Reply:
>In either case, adding Scheme to the policy is always a wise and safe thing, even if there is a bit of duplication!

The schem is already included in the Appraisal ID.  there is no point in adding it again for each policy. It does not make things in any way safer.